File: word.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

const (
	NO_MATCH = iota
	YELLOW_MATCH
	GREEN_MATCH
)

// 0 no match
// 1 yellow match
// 2 green match

//grape 00101
//grape 00101 navel 11010

// letter is not at position; from a yellow or gray match
type notAt struct {
	letter   byte
	position int
}

func (n notAt) passes(word string) bool {
	return word[n.position] != n.letter
}

// how many times letter appears in the word, based on the yellow and green
// matches for it in a single guess. if the guess also had a gray match for
// the letter, the word has exactly that many, otherwise at least that many.
type letterCount struct {
	letter byte
	count  int
	exact  bool
}

func (c letterCount) passes(word string) bool {
	n := strings.Count(word, string(c.letter))
	if c.exact {
		return n == c.count
	}
	return n >= c.count
}

// 2 green match
type green struct {
	letter  byte
	poition int
}

func (g green) passes(word string) bool {
	return word[g.poition] == g.letter
}

type rule interface {
	passes(word string) bool // returns true if word passes rule
}

type rules []rule

func (rs rules) passes(word string) bool {
	for _, r := range rs {
		if !r.passes(word) {
			return false
		}
	}
	return true
}

type arg struct {
	letters string
	numbers string
}

func (a arg) toRules() []rule {
	r := make([]rule, 5)
	counts := make(map[byte]int)
	grays := make(map[byte]bool)
	for i := range a.letters {
		letter := a.letters[i]
		switch int(a.numbers[i]) - '0' {
		case NO_MATCH:
			r[i] = notAt{letter, i}
			grays[letter] = true
		case YELLOW_MATCH:
			r[i] = notAt{letter, i}
			counts[letter]++
		case GREEN_MATCH:
			r[i] = green{letter, i}
			counts[letter]++
		}
	}
	for letter, count := range counts {
		r = append(r, letterCount{letter, count, grays[letter]})
	}
	for letter := range grays {
		if counts[letter] == 0 {
			r = append(r, letterCount{letter, 0, true})
		}
	}
	return r
}

func buildRules(args []arg) rules {
	rules := make([]rule, 0, len(args)*5)
	for _, arg := range args {
		rules = append(rules, arg.toRules()...)
	}
	return rules
}

func buildArgs(a []string) []arg {
	n := len(a) / 2

	args := make([]arg, n)

	for i := 0; i < n; i++ {
		args[i].letters = a[2*i]
		args[i].numbers = a[2*i+1]
	}
	return args
}

func word() {
	rules := buildRules(buildArgs(os.Args[1:]))
	r := bufio.NewReader(os.Stdin)
	line, _, err := r.ReadLine()
	for err == nil {
		if rules.passes(string(line)) {
			fmt.Println(string(line))
		}
		line, _, err = r.ReadLine()
	}
}

File: site/web/js/word.js
// Filtering and ranking, mirroring the Go program in legacy/. A guess is a 5 letter word
// and a result of 5 digits: 0 gray, 1 yellow, 2 green, e.g. crane 01020.

const NO_MATCH = 0
const YELLOW_MATCH = 1
const GREEN_MATCH = 2

// letter is not at position; from a yellow or gray match
const notAt = (letter, position) => word => word[position] !== letter

// letter is at position; from a green match
const green = (letter, position) => word => word[position] === letter

// how many times letter appears in the word, based on the yellow and green
// matches for it in a single guess. if the guess also had a gray match for
// the letter, the word has exactly that many, otherwise at least that many.
const letterCount = (letter, count, exact) => word => {
    const n = word.split(letter).length - 1
    return exact ? n === count : n >= count
}

function toRules(letters, numbers) {
    const rules = []
    const counts = {}
    const grays = {}
    for (let i = 0; i < letters.length; i++) {
        const letter = letters[i]
        switch (Number(numbers[i])) {
            case NO_MATCH:
                rules.push(notAt(letter, i))
                grays[letter] = true
                break
            case YELLOW_MATCH:
                rules.push(notAt(letter, i))
                counts[letter] = (counts[letter] || 0) + 1
                break
            case GREEN_MATCH:
                rules.push(green(letter, i))
                counts[letter] = (counts[letter] || 0) + 1
                break
        }
    }
    for (const letter in counts) {
        rules.push(letterCount(letter, counts[letter], !!grays[letter]))
    }
    for (const letter in grays) {
        if (!counts[letter]) {
            rules.push(letterCount(letter, 0, true))
        }
    }
    return rules
}

// guesses is a list of { letters, numbers }
function buildRules(guesses) {
    return guesses.flatMap(g => toRules(g.letters, g.numbers))
}

const passes = (rules, word) => rules.every(rule => rule(word))

// returns an error message, or "" if the guess and result are valid
function validateGuess(letters, numbers) {
    if (!/^[a-z]{5}$/.test(letters)) {
        return `guess "${letters}" must be 5 letters`
    }
    if (!/^[012]{5}$/.test(numbers)) {
        return `result "${numbers}" for guess "${letters}" must be 5 digits of 0, 1 or 2`
    }
    return ""
}

// counts each letter once per word, so counts are the number of words that
// contain the letter. anything other than a-z is ignored.
function countWord(counts, word) {
    const seen = {}
    for (const letter of word) {
        if (letter in counts && !seen[letter]) {
            seen[letter] = true
            counts[letter]++
        }
    }
}

// letters ordered from most to fewest words containing them
function letterOrder(words) {
    const counts = {}
    for (let i = 0; i < 26; i++) {
        counts[String.fromCharCode(97 + i)] = 0
    }
    for (const word of words) {
        countWord(counts, word)
    }
    return Object.keys(counts).sort((a, b) => counts[b] - counts[a])
}

// negative if first is better than second: whichever has the most common letter
// the other lacks wins. same letters (e.g. anagrams like least and slate), fall
// back to alphabetical
function cmpWords(order, first, second) {
    for (const letter of order) {
        const firstContains = first.includes(letter)
        const secondContains = second.includes(letter)
        if (firstContains && !secondContains) {
            return -1
        } else if (secondContains && !firstContains) {
            return 1
        }
    }
    return first < second ? -1 : first > second ? 1 : 0
}

// the words that fit every guess, best first
function rankWords(guesses, wordList) {
    const rules = buildRules(guesses)
    const words = wordList.filter(word => passes(rules, word))
    const order = letterOrder(words)
    return words.sort((a, b) => cmpWords(order, a, b))
}
